host/sgx/linux/vdso.c

```diff
@@ -0,0 +1,205 @@
+             * these registers until the function returns. Instead,
+             * we return zero to avoid dealing with such complexities
+             * and also to use similar implementation as regular enter. */
+            return_args->rdi = arg1;
+            return_args->rsi = arg2;
+            result = 0;
+            break;
+        }
+        case ENCLU_ERESUME:
+        {
+            /* Hardware exceptions occur */
+
+            oe_host_exception_context_t host_context = {0};
+
+            host_context.rax = ENCLU_ERESUME;
+            host_context.rbx = run->tcs;
+
+            /* AEP is assigned by vDSO implementation */
+
+            oe_host_handle_exception(&host_context);
+
+            result = ENCLU_ERESUME;
+            break;
+        }
+    }
+
+done:
+    /* If the result <= 0, the value will be forwared as the return
+     * value of _vdso_sgx_enter_enclave. Otherwise, _vdso_sgx_enter_enclave
+     * will invoke the ENCLU[result] instead of returning to the caller. */
+    return result;
+}
+
+oe_result_t oe_vdso_enter(
+    void* tcs,
+    uint64_t arg1,
+    uint64_t arg2,
+    uint64_t* arg3,
+    uint64_t* arg4,
+    oe_enclave_t* enclave,
+    oe_ecall_context_t* ecall_context)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    struct sgx_enclave_run run = {0};
+    vdso_args_t return_args = {0};
+    int return_value = 0;
+    uint32_t mxcsr = 0;
+    uint16_t fcw = 0;
+
+    run.tcs = (uint64_t)tcs;
+    run.user_handler = (uint64_t)oe_vdso_user_handler;
+    run.user_data = (uint64_t)&return_args;
+
+    while (1)
+    {
+        // Compiler will usually handle this on exiting a function that uses
+        // AVX, but we need to avoid the AVX-SSE transition penalty here
+        // manually as part of the transition to enclave. See
+        // https://software.intel.com/content/www/us/en/develop/articles
+        // /avoiding-avx-sse-transition-penalties.html
+        if (oe_is_avx_enabled)
+            OE_VZEROUPPER;
+
+        asm volatile("stmxcsr %[mxcsr] \n\t" // Save MXCSR
+                     "fstcw %[fcw] \n\t"     // Save x87 control word
+                     :
+                     : [fcw] "m"(fcw), [mxcsr] "m"(mxcsr)
+                     :);
+
+        return_value = (_vdso_sgx_enter_enclave)(
+            arg1,
+            arg2,
+            (uint64_t)ecall_context,
+            ENCLU_EENTER,
+            0 /* r8 */,
+            0 /* r9 */,
+            &run);
+
+        if (return_value < 0)
+            OE_RAISE(OE_FAILURE);
+
+        asm volatile("fldcw %[fcw] \n\t"     // Restore x87 control word
+                     "ldmxcsr %[mxcsr] \n\t" // Restore MXCSR
+                     :
+                     : [fcw] "m"(fcw), [mxcsr] "m"(mxcsr)
+                     :);
+
+        arg1 = return_args.rdi;
+        arg2 = return_args.rsi;
+
+        // Make an OCALL if needed.
+        oe_code_t code = oe_get_code_from_call_arg1(arg1);
+        if (code == OE_CODE_OCALL)
+        {
```

Comment:
since stack already unwound from vdso_enter_enclave, not sure if you will see correct stack trace for nested ecall->ocall->ecall in gdb. you may be OK if stitching happened above this frame.

Reply:
Thanks for pointing this out. I'll double check.

Reply:
OE doesn't support reentrant ecalls btw (ecall -> ocall -> ecall).